scripts/linux/nestedAgentLock.sh

```diff
@@ -0,0 +1,243 @@
+            agentsAllLockedCorrectly=false
+            break
+        fi
+    done
+
+    echo $agentsAllLockedCorrectly
+}
+
+function unlock_agents() {
+    agents=("$@")
+
+    for agentId in "${agents[@]}"; do
+        agentCapabilities=$(curl -s -u :$PAT --request GET "https://dev.azure.com/msazure/_apis/distributedtask/pools/$POOL_ID/agents/$agentId?includeCapabilities=true&api-version=$API_VER")
+        lockStatus=$(echo $agentCapabilities | jq '.userCapabilities | .status')
+
+        if [ $lockStatus = '"unlocked_$(Build.BuildId)"' ]; then
+            echo "Unlocking agent $agentId"
+
+            newAgentUserCapabilities=$(echo $agentCapabilities | jq '.userCapabilities | (.["status"]) |= "unlocked"')
+            update_capabilities "$agentId" "$newAgentUserCapabilities"
+        fi
+    done
+}
+
+process_args $@
+
+# Install pre-requisite 'jq'
+CMD=jq
+echo "Validating jq"
+if ! command -v $CMD &>/dev/null; then
+    echo "Command '$CMD' not found, Installing '$CMD'"
+    sudo add-apt-repository universe
+    sudo apt-get update
+    sudo apt-get install -y $CMD
+fi
+echo "Done validating jq"
+
+startSeconds=$((SECONDS))
+endSeconds=$((SECONDS + $TIMEOUT_SECONDS))
+backoffSeconds=1
+while true && [ $((SECONDS)) -lt $endSeconds ]; do
+    # Wait 1-10 seconds to retry locking agents.
+    # Random delay to avoid multiple instances of the script thrashing.
+    sleep $[ ( $RANDOM % 10 )  + $backoffSeconds ]s
+    backoffSeconds=$(($backoffSeconds * 2))
+
+    echo "Attempting to lock $AGENTS_NEEDED agents from the agent group $AGENT_GROUP..."
+    agentsInfo=$(curl -s -u :$PAT --request GET "https://dev.azure.com/msazure/_apis/distributedtask/pools/$POOL_ID/agents?includeCapabilities=true&api-version=$API_VER")
+    unlockedAgents=($(echo $agentsInfo | jq '.value | .[] | select(.userCapabilities.status=="unlocked" and .userCapabilities."agent-group"=='\"$AGENT_GROUP\"') | .id' | tr -d '[], "'))
+
+    echo "Found these unlocked agents:"
+    echo ${unlockedAgents[@]}
+
+    if [ ${#unlockedAgents[*]} -ge $AGENTS_NEEDED ]; then
+        # If we have enough agents, get random agents and book them all.
+        shuffledUnlockedAgents=($(shuf -e "${unlockedAgents[@]}"))
+        filteredAgents=(${shuffledUnlockedAgents[@]:0:$AGENTS_NEEDED})
+
+        echo "Locking these agents:"
+        echo ${filteredAgents[@]}
+
+        agentsAllLockedCorrectly=$(attempt_agent_lock "${filteredAgents[@]}")
+
+        # If something went wrong and we don't have all the agents locked, release the ones we still have booked.
+        # Else set the agent names as output to be used in the pipeline downstream.
+        if [ $agentsAllLockedCorrectly = false ]; then
+            echo "Conflicting agent lock detected. Unlocking all booked agents made here."
+            unlock_agents "${filteredAgents[@]}"
+
+        else
+            echo "Successfully locked agents"
+            set_devops_output_vars "${filteredAgents[@]}"
+            exit 0
```

Comment:
Do you want `exit` or `return` here? maybe you really want a `return` and not an exit -- which terminate the bash process?

Reply:
I was thinking we want to exit because we successfully locked the agents. Therefore the script should exit.

Reply:
Sure.